src/env.cc

```diff
@@ -2260,4 +2271,97 @@ v8::CpuProfile* Environment::StopCpuProfile(v8::ProfilerId profile_id) {
   return profile;
 }
 
+Mutex Environment::notifications_mutex_;
+uint64_t Environment::next_notification_id_(0);
+std::unordered_map<uint64_t, Environment*> Environment::notifications_;
+
+void Environment::RegisterNotification(
+    const v8::FunctionCallbackInfo<v8::Value>& args) {
+  CHECK(args[0]->IsFunction());
+
+  Mutex::ScopedLock lock(notifications_mutex_);
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  uint64_t id = ++Environment::next_notification_id_;
+  notifications_.emplace(id, env);
+
+  v8::Global<v8::Function> global;
+  global.Reset(isolate, args[0].As<v8::Function>());
+  env->notifications_callbacks_.emplace(id, std::move(global));
+
+  args.GetReturnValue().Set(v8::BigInt::New(isolate, id));
+}
+
+void Environment::UnregisterNotification(
+    const v8::FunctionCallbackInfo<v8::Value>& args) {
+  CHECK(args[0]->IsBigInt());
+
+  bool lossless = false;
+  uint64_t id = args[0].As<v8::BigInt>()->Uint64Value(&lossless);
+
+  if (!lossless) {
+    std::cout << "LOSSLESS TRUE\n";
+    return THROW_ERR_INVALID_NOTIFICATION(
+        args.GetIsolate(), "Invalid notification %un", id);
+  }
+
+  Mutex::ScopedLock lock(notifications_mutex_);
+  Environment* env = Environment::GetCurrent(args);
+
+  if (env->notifications_callbacks_.erase(id) == 0) {
+    return THROW_ERR_INVALID_NOTIFICATION(
+        args.GetIsolate(), "Invalid notification %llun", id);
+  }
+
+  env->notifications_queue_.erase(id);
+  notifications_.erase(id);
+}
+
+void Environment::GetNotifications(
+    const v8::FunctionCallbackInfo<v8::Value>& args) {
+  Mutex::ScopedLock lock(notifications_mutex_);
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  v8::LocalVector<Value> ids(isolate);
+  for (auto& pairs : env->notifications_callbacks_) {
+    ids.push_back(v8::BigInt::New(isolate, pairs.first));
+  }
```

Comment:
There's a relatively newish signature for `Array::New(...)` that takes an std::function as input to create each item. It's useful for avoiding having to use an intermediate variable like `ids` here, which carries it's own cost.

Reply:
yep, needs to be benchmarked carefully. Mentioned it only because it's an option :-)

Reply:
Nevermind. Found.